### ISSUE TITLE
[Proposal] Dry-run support for artisan schedule:run

### ISSUE THREAD

**Issue by thejettdurham** [laravel/internals#196]
I'm working on a project now with a fairly complex schedule logic for many long-running tasks in the Console Kernel. In this case, it would be helpful to have a way to simply validate the scheduling logic without actually kicking off any of the tasks.

Adding an option to `artisan schedule:run` seems to be the most straight-forward approach. For example: `artisan schedule:run --dry` would list the commands that _would_ run according to the schedule, but wouldn't actually run the command(s).

If there are no objections to this proposal and/or my suggested approach, I already have a PR ready for submission :)

**Comment by tomschlick** [comment]
Looks like a good submission to me.

Just curious, how often do you see yourself needing to dry run things? Is it just for testing / debugging or is there some use case I'm not considering?

**Comment by bkuhl** [comment]
Seems like it would have limited usefulness since you'd have to run the command at the very minute it was scheduled.  More clear scheduling functionality/methods I think would be a better solution.

**Comment by thejettdurham** [comment]
Testing and debugging is the primary use case of the proposal. Here are a few cases that I've encountered in my app that led me to hacking up a dry-run option in the schedule:run command:

- Validating that there are no issues with executing `artisan schedule:run` across different environments (dev, prod, etc). For example, a constructor called during autoloading may call out to a database available to dev, but hasn't yet been made available in prod.
- Testing one or more `when` and/or `skip` closures on a command.
- Testing imperative logic that adds `$schedule` calls non-declaratively. For example: conditionally adding a command for each element in an array depending on the value.
- Being able to prove to my boss that the scheduler will run X commands at Y time and Z external condition(s) without actually running the commands and potentially affecting live data.

True, most simple scheduler setups have no need for dry-run, but since Laravel affords such flexibility in triggering commands based on external conditions in addition to the current time, I've found it useful to test only the schedule logic itself. This is especially true in my case where I'm scheduling commands that were written by other members of the team and can potentially take a long time to run, blocking the rest of the scheduled tasks from starting.
